eventauth.go

```diff
@@ -204,25 +191,25 @@ func Allowed(event Event, authEvents AuthEvents) error {
 // createEventAllowed checks whether the m.room.create event is allowed.
 // It returns an error if the event is not allowed.
 func createEventAllowed(event Event) error {
-	if event.StateKey == nil {
+	if !event.IsState() {
 		return errorf("create event missing state key")
 	}
-	if *event.StateKey != "" {
-		return errorf("create event state key is not empty: %q", event.StateKey)
+	if event.StateKey() != "" {
```

Comment:
I feel really uncomfortable with this given the default semantics if there is no state key. We're relying completely that we have checked `IsState()` prior to calling this: if we (re)move that then this check will NOT fail. This is really subtle: it's clearer in the previous diff because it's obvious there is a correlation since you're doing explicit `nil` checks on the same field.

I'd rather we removed `IsState()` and only exposed a nullable string return value for `StateKey()` to make this completely obvious.